sections/testing/unit-testing.tex

```diff
@@ -0,0 +1,97 @@
+For the frontend of the system, it would be possible to test whether each component is rendered to the user, if all the information is correct, what happens if you use 0 as input to the function, 1 as input, 3 as input.. And so on.
+Instead, we chose to focus on testing if the components that are conditionally rendered are shown when the conditions are fulfilled.
+An example of this could be the \textit{Show User} page, where the system should show tutor info if and only if the user is actually a tutor.
+This now leads to test cases like is the info rendered if:
+\begin{itemize}
+    \item the user is a tutor?
+    \item the user is a tutor?
+    \item the user is a tutor, but the tutor information is invalid?
+    \item the user is a tutor, but the tutor information is incomplete?
+\end{itemize}
+
+In addition to the conditional rendering, any business logic located on the frontend should be tested to ensure both that it works as intended, but also to ensure that any future updates to the code will not break it, which is referred to as regression testing.
+
+For the API, some functionality like the ORM is handled by external packages. Since this is not code we're writing ourselves, we have chosen to assume that the developers of the packages have tested the software and we will not have to.
+We focus on unit testing our controllers and middleware to ensure that they work as expected, and instead do some integration testing for the external packages, to ensure that they work as expected when integrated into our system. This will be further described in \autoref{sec:integrationTesting}.
+
+\subsection{Stubbing}
+A major obstacle for generating our tests was to avoid the calls to the ORM in the API, which would connect to the database.
+At first, we considered different ways to handle this connection, such as having a test database that would be wiped before each test to generate a controlled environment.
+However, after doing more research we realized that instead of connecting to a real world configuration, it would be beneficial to be able to control exactly which values should be returned.
+This principle is called stubbing \cite{SoftwareTesting}.
+The idea is that instead of calling a given function, we create a stub function that is called instead.
+An example of this can be seen in the unit test in \autoref{lst:unitTestStubbing}, where we stub the function call to \texttt{getByEmail} in the \texttt{userService}.
+In the real code, the code seen on \autoref{lst:getByEmail} would be called as a part of the call to the login function on line 16, which connects to the database, but instead of doing so we use the Sinon library to stub the function, and tell it to instead throw an exception, so we can see how the controller handles this case.
```

Comment:
```suggestion
In the real code, the code seen on \autoref{lst:getByEmail} would be called as a part of the call to the login function on line 16, which connects to the database, but instead of doing so we use the Sinon library to stub the function, and tell it to throw an exception instead, so we can see how the controller handles this case.
```